### PR TITLE
🌱 Add e2e test for node drain timeout feature

### PR DESCRIPTION
**Pull request by namnx228**
**What this PR does / why we need it**:
This is an e2e test for the node drain timeout feature implemented in PR #3662 
This test aims to make sure that all machines can be removed after a specific amount of time even when they have workloads running inside that block the draining process. 
The test scenario is as follows:
- Create a management cluster with one control plane and one worker machine.
- Update the `NodeDrainTImeout` field of the `MachineDeployment`, and wait until all worker machines to be updated.
- Add some workloads that cannot be evicted during the node draining process (using pod deployment and `podDisruptionBudget`)
- Make sure that all worker nodes can still be deleted after a specific amount of time.

- Untaint the control plane so they can run workloads.
- Update the `NodeDrainTImeout` field of KCP and all the existing control plane machine. 
- Scale up the KCP and make sure all new control plane machines have the new value of `NodeDrainTimeout`
- Wait until all control plane machines to have workloads running inside them.
- Make sure that all control plane nodes can still be deleted after a specific amount of time.

### CODE REVIEW

File: test/framework/machines.go [hunk 1]
```diff
@@ -22,11 +22,51 @@ import (
...
+	MachineDeployments         []*clusterv1.MachineDeployment
+	NodeDrainTimeout           *metav1.Duration
+	ClusterProxy               ClusterProxy
+	Cluster                    *clusterv1.Cluster
+	WaitForMachinesToBeUpdated []interface{}
+}
+
+//UpdateNodeDrainTimeoutInMachineDeployment updates the nodeDrainTimeout field in the machinedeployment
+//and wiat until all machines are updated
+func UpdateNodeDrainTimeoutInMachineDeployment(ctx context.Context, input UpdateNodeDrainTimeoutInMachineDeploymentInput) {
```
Comment (sedefsavas) [inline test/framework/machines.go:44]: Instead of patching the node drain timeout here, how about having a new flavor with node drain timeout?
Looks like at the beginning of the node drain test, timeout is patched right away so we can have a cluster-template-nodedrain directory under `data/infrastructure-docker` that will be added to the cluster-template using `kustomize`.

File: test/framework/deployment_helpers.go [hunk 2]
```diff
@@ -247,3 +282,128 @@ func WaitForDNSUpgrade(ctx context.Context, input WaitForDNSUpgradeInput, interv
 		return false, nil
 	}, intervals...).Should(BeTrue())
 }
+
+type AddUnevictablePodInput struct {
+	ClusterProxy                       ClusterProxy
```
Comment (sedefsavas) [inline test/framework/deployment_helpers.go:287]: We can pass the workload cluster proxy instead of passing the management cluster proxy, because AddUnevictablePod only works on workload cluster.

So from the node_drain_timeout.go, workload cluster can be reached as follows:

```
		// Get a ClusterBroker so we can interact with the workload cluster
		selfHostedClusterProxy = input.BootstrapClusterProxy.GetWorkloadCluster(context.TODO(), cluster.Namespace, cluster.Name)

```

After this passing workload cluster client as ClusterProxy to this function as input should work.

File: test/e2e/node_drain_timeout.go [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+				Cluster:                   cluster,
+				MachineDeployment:         md,
+				WaitForMachineDeployments: nodeDrainTimeoutMachineDeploymentInterval,
+				Replicas:                  0,
+			})
+		}
+
+		By("Update nodeDrainTimeout field of the existing and new controlplane machines.")
+		numScaledUpControlPlane := 3
+		framework.UpdateControlplaneNodeDrainTimeout(ctx, framework.UpdateControlplaneNodeDrainTimeoutInput{
```
Comment (sedefsavas) [inline test/e2e/node_drain_timeout.go:139]: Node drain timeout for control plane and machine deployment can be embedded into cluster-template node-drain flavor so that we do not have to patch them.

File: test/framework/controlplane_helpers.go [hunk 4]
```diff
@@ -351,3 +357,185 @@ func controlPlaneMachineOptions() []client.ListOption {
...
+	})
+	for i := range controlplaneMachines {
+		patchHelper, err := patch.NewHelper(&controlplaneMachines[i], mgmtClient)
+		Expect(err).ToNot(HaveOccurred())
+		controlplaneMachines[i].Spec.NodeDrainTimeout = input.NodeDrainTimeout
+		Expect(patchHelper.Patch(context.TODO(), &controlplaneMachines[i])).To(Succeed())
+	}
+
+	log.Logf("Add new controlplane machine to see if new machines get new NodeDrainTimeout value")
+	ScaleAndWaitControlPlane(ctx, ScaleAndWaitControlPlaneInput{
```
Comment (sedefsavas) [inline test/framework/controlplane_helpers.go:401]: I don't see much benefit of adding the timeout and scaling up controlPlane compared to having node drain value in cluster-template and having 3 control plane machines. These two will go through scale up. 
We can remove `ScaleAndWaitControlPlane` and `UpdateControlplaneNodeDrainTimeout` if we use a custom template flavor.

File: test/e2e/node_drain_timeout.go [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+		framework.UpdateNodeDrainTimeoutInMachineDeployment(ctx, framework.UpdateNodeDrainTimeoutInMachineDeploymentInput{
+			ClusterProxy:               input.BootstrapClusterProxy,
+			Cluster:                    cluster,
+			MachineDeployments:         machineDeployments,
+			NodeDrainTimeout:           nodeDrainTimeoutDuration,
+			WaitForMachinesToBeUpdated: input.E2EConfig.GetIntervals(specName, "wait-machine-updated"),
+		})
+
+		By("Add a deployment and podDisruptionBudget to the workload cluster. The deployed pods cannot be evicted in the node draining process.")
+		framework.AddUnevictablePod(ctx, framework.AddUnevictablePodInput{
```
Comment (fabriziopandini) [inline test/e2e/node_drain_timeout.go:119]: Might be
```suggestion
		framework.DeployUnevictablePod(ctx, framework.AddUnevictablePodInput{
```

**Review by fabriziopandini — COMMENTED** [review]
First pass to the code
Also, I would like to take a look at the describe test sequence and see if it can be simplified, by focusing on the essential part:

> Create a management cluster with one control plane and one worker machine.
>  Update the NodeDrainTImeout field of the MachineDeployment, and wait until all worker machines to be updated.

Why not set the NodeDrainTImeout field in the template, so we are will be ready to test drain without being forced to wait for a rollout (now that cluster templates are generated with kustomize, this shouldn't be so complex to achieve, let me know if you need some guidance).

> Add some workloads that cannot be evicted during the node draining process (using pod deployment and podDisruptionBudget)
> Make sure that all worker nodes can still be deleted after a specific amount of time.

ok

> Untaint the control plane so they can run workloads.

forgive my poor memory, but is this something we want to support?

> Update the NodeDrainTImeout field of KCP and all the existing control plane machine.
> Scale up the KCP and make sure all new control plane machines have the new value of NodeDrainTimeout

Why not to set the NodeDrainTImeout field in the template, so we are will be ready to test drain without being forced to wait KCP rollout.

> Wait until all control plane machines to have workloads running inside them.
> Make sure that all control plane nodes can still be deleted after a specific amount of time.

File: test/e2e/node_drain_timeout.go [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+		framework.UpdateControlplaneNodeDrainTimeout(ctx, framework.UpdateControlplaneNodeDrainTimeoutInput{
+			Controlplane:               controlplane,
+			NodeDrainTimeout:           nodeDrainTimeoutDuration,
+			ClusterProxy:               input.BootstrapClusterProxy,
+			Cluster:                    cluster,
+			ScaleUpTo:                  int32(numScaledUpControlPlane),
+			WaitForMachinesToBeUpdated: input.E2EConfig.GetIntervals(specName, "wait-controlplane-updated"),
+		})
+		By("Deploy workload on the master node. The workload is actually the pods that we deployed above.")
+		framework.DeployWorkloadOnControlplaneNode(ctx, framework.DeployWorkloadOnControlplaneNodeInput{
```
Comment (fabriziopandini) [inline test/e2e/node_drain_timeout.go:148]: Should this be `DeployUnevictablePod` with a Toleration and a NodeSelector?

**Comment by namnx228** [comment]
Hi @sedefsavas and @fabriziopandini,
Thanks for your reviews and suggestions,
I agree that we should have NodeDrainTimeout in the cluster-template before creating the cluster. However, I have no idea how to make the timeout as a variable in the cluster-template, and then pass the value from the test code? We can do it with some other variables like CLUSTER_NAME or WORKER_MACINE_COUNT, but it is not a good idea to add NODE_DRAIN_TIMEOUT as a new variable to ClusterAPI just for this test purpose. Are there any workarounds for this?

**Comment by fabriziopandini** [comment]
@namnx228 
you can add your variable in the test e2e configuration 
https://github.com/kubernetes-sigs/cluster-api/blob/be5ff75585e4762ae8ef2a9166dcfd16a9f2b8a1/test/e2e/config/docker.yaml#L76-L88

and then create a new template under https://github.com/kubernetes-sigs/cluster-api/tree/master/test/e2e/data/infrastructure-docker (put it in new folder named cluster-template-drain, add your kustomize file + your kustomize patches there), and then add you template to the list of templates to be generated in https://github.com/kubernetes-sigs/cluster-api/blob/be5ff75585e4762ae8ef2a9166dcfd16a9f2b8a1/test/e2e/Makefile#L57-L62

**Comment by namnx228** [comment]
@fabriziopandini Thanks for your detailed instruction.
I have another question: How to pass the NodeDrainTimeout variable value the docker.yaml to the cluster template?
I know how to get the value from docker.yaml by using function GetVariables() like this: https://github.com/kubernetes-sigs/cluster-api/blob/169440de58bb676909af1e400f8cede84d0120fb/test/e2e/quick_start.go#L85, but to apply it to create a cluster, we need to pass NodeDrainTimeout as a parameter to the function `ApplyClusterTemplateAndWait`. And this function takes cluster configuration input from `ConfigClusterInput`. However, all the variables in this struct are parameters that we can pass to Clusterctl. That means to pass the NodeDrainTimeout to the cluster template, we need to make some changes on Clusterctl site to add an extra parameter for `NodeDrainTimeout`. I am not sure if it is a good idea to add a new parameter to Clusterctl just to run some e2e tests. This is the problem that is blocking me now. 

test/framework/clusterctl/client.go 86 
 test/e2e/node_drain_timeout.go 80

test/framework/clusterctl/e2e_config.go 387

**Comment by namnx228** [comment]
> > Untaint the control plane so they can run workloads.
> 
> forgive my poor memory, but is this something we want to support?

Yes, we support the NodeDrainTimeout even for KCP. The reason is that we have some use cases where workload can also be deployed on controlplane nodes.

File: test/e2e/node_drain_timeout.go (hunk 6 above)
Comment (namnx228) [reply test/e2e/node_drain_timeout.go:148]: Good suggestion, I will make an update on it.

File: test/framework/deployment_helpers.go (hunk 2 above)
Comment (namnx228) [reply test/framework/deployment_helpers.go:287]: Thanks for your suggestion. I will make an update on it

**Comment by fabriziopandini** [comment]
> How to pass the NodeDrainTimeout variable value the docker.yaml to the cluster template?

Variable in docker.yaml are used to set env variables for the call to ApplyClusterTemplateAndWait, so you can use them from your template by leveraging on [variable substitution](https://cluster-api.sigs.k8s.io/clusterctl/provider-contract.html#variables) (simply using `${DRAIN_TIMEOUT}` in your template).

File: test/e2e/config/docker.yaml [hunk 7]
```diff
@@ -86,6 +86,9 @@ variables:
   EXP_CLUSTER_RESOURCE_SET: "true"
   EXP_MACHINE_POOL: "true"
 
+intervals:
+  default/wait-controllers: ["3m", "10s"]
```
[the remark this answers is not in the document]
Comment (namnx228) [reply test/e2e/config/docker.yaml]: Fixed! Thanks

File: test/e2e/node_drain_timeout_test.go [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+
+package e2e
+
+import (
+	"context"
+
+	. "github.com/onsi/ginkgo"
+)
+
+var _ = Describe("When a node cannot be drained in time and need to be removed forcefully", func() {
```
[the remark this answers is not in the document]
Comment (namnx228) [reply test/e2e/node_drain_timeout_test.go]: Fixed! Thanks

File: test/framework/controlplane_helpers.go [hunk 9]
```diff
@@ -21,12 +21,18 @@ import (
 
 	. "github.com/onsi/ginkgo"
 	. "github.com/onsi/gomega"
```
[the remark this answers is not in the document]
Comment (namnx228) [reply test/framework/controlplane_helpers.go:24]: Solved! Thanks

**Comment by namnx228** [comment]
All required changes have been made. Please take a look again @sedefsavas @fabriziopandini. Thanks

File: test/e2e/node_drain_timeout.go [hunk 10]
```diff
@@ -0,0 +1,168 @@
...
+
+		// Setup a Namespace where to host objects for this spec and create a watcher for the namespace events.
+		namespace, cancelWatches = setupSpecNamespace(ctx, specName, input.BootstrapClusterProxy, input.ArtifactFolder)
+
+	})
+
+	It("A note should be forcefully removed if it cannot be drained in time", func() {
+
+		By("Creating a workload cluster")
+		controlPlaneReplicas := 3
```
Comment (fabriziopandini) [inline test/e2e/node_drain_timeout.go:81]: Is it required to have 3 control plane replicas? 
I assume that or replica could be enough to verify drain timeout on KCP nodes...

File: test/e2e/node_drain_timeout.go [hunk 11]
```diff
@@ -0,0 +1,168 @@
...
+	return intervals
+}
+
+func convertMachineDeploymentDurationToInterval(duration *metav1.Duration) []interface{} {
+	return convertDurationToInterval(duration, 1)
+}
+
+func convertKCPDurationToInternal(duration *metav1.Duration, numKCP int) []interface{} {
+	return convertDurationToInterval(duration, numKCP-1)
+}
```
Comment (fabriziopandini) [inline test/e2e/node_drain_timeout.go:168]: Those funcs can be dropped if we move down to one control plane replica, and also convertDurationToInterval can be simplified by removing the `delayRate` parameter

File: test/e2e/node_drain_timeout.go [hunk 12]
```diff
@@ -0,0 +1,168 @@
...
+	})
+}
+
+func convertDurationToInterval(duration *metav1.Duration, delayRate int) []interface{} {
+	minIntervalDuration := duration.Duration
+	maxIntervalDuration := (duration.Duration + time.Minute*2) * time.Duration(delayRate)
+	intervals := make([]interface{}, 0, 2)
+	intervals = append(intervals, maxIntervalDuration.String(), minIntervalDuration.String())
+	return intervals
+}
```
Comment (fabriziopandini) [inline test/e2e/node_drain_timeout.go:160]: I'm a little bit confused by this method, because AFAIK the intervals are passed to Eventually which expects a timeout and a polling interval, while here you are computing two intervals and adding 0,2 const.

File: test/e2e/data/infrastructure-docker/bases/cluster-with-kcp.yaml [hunk 13]
```diff
@@ -51,6 +51,7 @@ metadata:
   labels:
     kcp-adoption.step2: ""
 spec:
+  nodeDrainTimeout: ${NODE_DRAIN_TIMEOUT}
```
Comment (fabriziopandini) [inline test/e2e/data/infrastructure-docker/bases/cluster-with-kcp.yaml:54]: If we are applying this to the base `cluster-with-kcp` this means that all the E2E tests based on the default template are going to interrupt drain after 60s. I'm not sure it is ok to change the behavior of all the tests... is there some specific reason why you went down this path instead of refining the usage of a dedicated template?

Comment (namnx228) [reply test/e2e/data/infrastructure-docker/bases/cluster-with-kcp.yaml:54]: Maybe I didn't get you right. If I understand right, you mentioned above that a small change should be only like a patch to the standard bases?

File: test/e2e/node_drain_timeout.go (hunk 10 above)
Comment (namnx228) [reply test/e2e/node_drain_timeout.go:81]: Is it possible to scale down the KCP to zero?

File: test/e2e/data/infrastructure-docker/bases/cluster-with-kcp.yaml (hunk 13 above)
Comment (namnx228) [reply test/e2e/data/infrastructure-docker/bases/cluster-with-kcp.yaml:54]: Oh I got your idea. Did you mean to redefine only KCP and machinedeployment to avoid duplicating the whole yaml files?

File: test/e2e/node_drain_timeout.go (hunk 12 above)
Comment (namnx228) [reply test/e2e/node_drain_timeout.go:160]: I misunderstand the polling interval. I will fix it. 
The reason for 2 minutes added to the interval because this timeout interval is calculated based on the NodeDrainTimeout. However, after NodeDrainTimeout is over, the cluster still needs more time to completely remove machines. That is why we need extra time for this process.

File: test/e2e/data/infrastructure-docker/bases/cluster-with-kcp.yaml (hunk 13 above)
Comment (fabriziopandini) [reply test/e2e/data/infrastructure-docker/bases/cluster-with-kcp.yaml:54]: Sorry if I wasn't clear.
My suggestion is to not change base yamls, but to create kustomize patches to change the base for the node drain test following the approach described [here](https://kubectl.docs.kubernetes.io/guides/introduction/kustomize/#2-create-variants-using-overlays). 

An example of a kustomize patch can be found in cluster-api at https://github.com/kubernetes-sigs/cluster-api/blob/ad28a70ab0c3ee804cebc8084fdbb92e10cc24b6/controlplane/kubeadm/config/manager/manager_image_patch.yaml (this patche changes the  image field, but you can use the same approach for adding nodeDrainTimeout)

In other words, IMO.
- it is fine to create a new template for node-drain
- this template should be created by composing existing bases and then applying overlay patches to add the nodeDrainTimeout field where required (instead of duplicating the whole bases)

Comment (namnx228) [reply test/e2e/data/infrastructure-docker/bases/cluster-with-kcp.yaml:54]: Got it. Thanks

File: test/e2e/node_drain_timeout.go [hunk 14]
```diff
@@ -0,0 +1,168 @@
...
+// NodeDrainTimeoutInput is the input for NodeDrainTimeoutSpec.
+type NodeDrainTimeoutSpecInput struct {
+	E2EConfig             *clusterctl.E2EConfig
+	ClusterctlConfigPath  string
+	BootstrapClusterProxy framework.ClusterProxy
+	ArtifactFolder        string
+	SkipCleanup           bool
+}
+
+// KCPUpgradeSpec implements a test that verifies KCP to properly upgrade a control plane with 3 machines.
```
[the remark this answers is not in the document]
Comment (namnx228) [reply test/e2e/node_drain_timeout.go:50]: Fix! Thanks

File: test/e2e/node_drain_timeout.go [hunk 15]
```diff
@@ -0,0 +1,168 @@
...
+		Expect(input.ClusterctlConfigPath).To(BeAnExistingFile(), "Invalid argument. input.ClusterctlConfigPath must be an existing file when calling %s spec", specName)
+		Expect(input.BootstrapClusterProxy).ToNot(BeNil(), "Invalid argument. input.BootstrapClusterProxy can't be nil when calling %s spec", specName)
+		Expect(os.MkdirAll(input.ArtifactFolder, 0755)).To(Succeed(), "Invalid argument. input.ArtifactFolder can't be created for %s spec", specName)
+
+		Expect(input.E2EConfig.GetIntervals(specName, "wait-deployment-available")).ToNot(BeNil())
+		Expect(input.E2EConfig.GetIntervals(specName, "wait-node-drain")).ToNot(BeNil())
+
+		// Setup a Namespace where to host objects for this spec and create a watcher for the namespace events.
+		namespace, cancelWatches = setupSpecNamespace(ctx, specName, input.BootstrapClusterProxy, input.ArtifactFolder)
+
```
[the remark this answers is not in the document]
Comment (namnx228) [reply test/e2e/node_drain_timeout.go:75]: Fixed! Thanks

File: test/e2e/node_drain_timeout.go [hunk 16]
```diff
@@ -0,0 +1,168 @@
...
+	BeforeEach(func() {
+		Expect(ctx).NotTo(BeNil(), "ctx is required for %s spec", specName)
+		input = inputGetter()
+		Expect(input.E2EConfig).ToNot(BeNil(), "Invalid argument. input.E2EConfig can't be nil when calling %s spec", specName)
+		Expect(input.ClusterctlConfigPath).To(BeAnExistingFile(), "Invalid argument. input.ClusterctlConfigPath must be an existing file when calling %s spec", specName)
+		Expect(input.BootstrapClusterProxy).ToNot(BeNil(), "Invalid argument. input.BootstrapClusterProxy can't be nil when calling %s spec", specName)
+		Expect(os.MkdirAll(input.ArtifactFolder, 0755)).To(Succeed(), "Invalid argument. input.ArtifactFolder can't be created for %s spec", specName)
+
+		Expect(input.E2EConfig.GetIntervals(specName, "wait-deployment-available")).ToNot(BeNil())
+		Expect(input.E2EConfig.GetIntervals(specName, "wait-node-drain")).ToNot(BeNil())
```
[the remark this answers is not in the document]
Comment (namnx228) [reply test/e2e/node_drain_timeout.go:71]: Fixed! Thanks

File: test/e2e/node_drain_timeout.go [hunk 17]
```diff
@@ -0,0 +1,168 @@
...
+
+		// Setup a Namespace where to host objects for this spec and create a watcher for the namespace events.
+		namespace, cancelWatches = setupSpecNamespace(ctx, specName, input.BootstrapClusterProxy, input.ArtifactFolder)
+
+	})
+
+	It("A note should be forcefully removed if it cannot be drained in time", func() {
+
```
[the remark this answers is not in the document]
Comment (namnx228) [reply test/e2e/node_drain_timeout.go:79]: Fixed! Thanks

File: test/e2e/node_drain_timeout.go [hunk 18]
```diff
@@ -0,0 +1,168 @@
...
+		Expect(input.E2EConfig.GetIntervals(specName, "wait-deployment-available")).ToNot(BeNil())
+		Expect(input.E2EConfig.GetIntervals(specName, "wait-node-drain")).ToNot(BeNil())
+
+		// Setup a Namespace where to host objects for this spec and create a watcher for the namespace events.
+		namespace, cancelWatches = setupSpecNamespace(ctx, specName, input.BootstrapClusterProxy, input.ArtifactFolder)
+
+	})
+
+	It("A note should be forcefully removed if it cannot be drained in time", func() {
```
[the remark this answers is not in the document]
Comment (namnx228) [reply test/e2e/node_drain_timeout.go:78]: Fixed! Thanks

File: test/e2e/config/docker.yaml [hunk 19]
```diff
@@ -96,3 +98,7 @@ intervals:
   default/wait-machine-upgrade: ["20m", "10s"]
   default/wait-machine-pool-upgrade: ["5m", "10s"]
   default/wait-machine-remediation: ["5m", "10s"]
+  default/wait-deployment-available: ["3m", "10s"]
+  default/wait-node-drain: ["4m", "60s"]
+  default/wait-machine-updated: ["4m", "10s"]
+  default/wait-controlplane-updated: ["10m", "10s"]
```
[the remark this answers is not in the document]
Comment (namnx228) [reply test/e2e/config/docker.yaml:104]: Fixed! Thanks

**Comment by namnx228** [comment]
@fabriziopandini your change requests have been made. Please take a look again. Thanks.
Also, I have tested and seen that we cannot scale down to 0 KCP. There is always one remaining controlplane machine.  Therefore, the only way to test the drain on KCP nodes is to have 3 replicas, then scaling down to 1 replicas.

**Review by fabriziopandini — COMMENTED** [review]
@namnx228 now the template generation looks much cleaner, thanks!
Some small nits/cleanup to the helpers func, then this one is ready to go for me

File: test/e2e/node_drain_timeout.go (hunk 6 above)
Comment (namnx228) [reply test/e2e/node_drain_timeout.go:124]: Sorry, I don't get your point here. In `DeployWorkloadOnControlplaneNode` I untainted the master nodes and used NodeSelector to make sure that pods are deployed on all 3 master nodes. Was that what you mean? Or do you mean the description in `By(...)` should be updated?

File: test/framework/deployment_helpers.go [hunk 20]
```diff
@@ -69,6 +76,38 @@ func WaitForDeploymentsAvailable(ctx context.Context, input WaitForDeploymentsAv
...
 
+type WaitForDeploymentsAvailableClientsetInput struct {
+	ClientSet                           *kubernetes.Clientset
+	Deployment                          *appsv1.Deployment
+	WaitForDeploymentsAvailableInterval []interface{}
+	Context                             context.Context
+	Replicas                            int
+}
+
+func WaitForDeploymentsAvailableClientset(input WaitForDeploymentsAvailableClientsetInput) {
```
[the remark this answers is not in the document]
Comment (namnx228) [reply test/framework/deployment_helpers.go:87]: Unfortunately, we cannot change like this because there is already a function named `WaitForDeploymentsAvailable`

File: test/e2e/node_drain_timeout.go (hunk 6 above)
Comment (fabriziopandini) [reply test/e2e/node_drain_timeout.go:125]: The point here is that DeployUnevictablePod and DeployWorkloadOnControlplaneNode are doing the same job, which is to beploy an unevictable pod; the only difference is that they targets different nodes.
So IMO we should have a single method, DeployUnevictablePod, and add to this method new input parameters (a flag or bare tolerations/selectors) to support deployment on workers and on control-plane nodes

File: test/framework/deployment_helpers.go (hunk 20 above)
[the remark this answers is not in the document]
Comment (fabriziopandini) [reply test/framework/deployment_helpers.go:87]: Might be I'm missing something then, why we need a new method to solve the same problem?

[the remark this answers is not in the document]
Comment (namnx228) [reply test/framework/deployment_helpers.go:87]: The existing function uses ClusterProxy to access the workload cluster. However, ClusterProxy cannot deploy `podDisruptionBudge`. I don't remember clearly what is the error message, but it seems like a problem in the library. That is why I need to use ClientSet to access the workload cluster and deploy `podDisruptionBudge`. That is the different between the existing and the new function that I wrote.

[the remark this answers is not in the document]
Comment (fabriziopandini) [reply test/framework/deployment_helpers.go:87]: If I look at the code below, this func is just waiting, and this is exactly the same as what WaitForDeploymentsAvailable is doing (no need of using clientset for that, right?).

Instead, I'm expecting that the responsibility of deploying `podDisruptionBudget` is assigned to `DeployUnevictablePod`, do you agree?

[the remark this answers is not in the document]
Comment (namnx228) [reply test/framework/deployment_helpers.go:86]: Agree. I have fixed it to use the existing `WaitForDeploymentsAvailable` function.

File: test/e2e/node_drain_timeout.go (hunk 6 above)
Comment (namnx228) [reply test/e2e/node_drain_timeout.go]: I see.  Actually`DeployWorkloadOnControlplaneNode` was needed to untainted the controlplane nodes and then all exiting deployments are redeployed in controlplane nodes. The existing deployments were deployed in previous steps in worker nodes to test the node drain timeout for the `machinedeployment`. After all worker machines have been deleted, these deployments  are waiting until there are available nodes to redeploy them. 
In the newest fix, I have deleted `DeployWorkloadOnControlplaneNode` and moved its functionality into `DeployUnevictablePod`. So now we only one function for deploying unevictable pods.

**Comment by namnx228** [comment]
Hi @fabriziopandini @sedefsavas, all the required changed have been made. Please take a look again. Thanks.

**Comment by namnx228** [comment]
At the moment, I have multiple commits. I hope to push a separated commit for the new fix can help reviewers to easier see if the change requests have been made well. I will squash the commits eventually.

File: test/framework/deployment_helpers.go [hunk 21]
```diff
@@ -245,3 +283,139 @@ func WaitForDNSUpgrade(ctx context.Context, input WaitForDNSUpgradeInput, interv
...
+	ControlPlane                       *controlplanev1.KubeadmControlPlane
+	IsDeploymentExisted                bool
+	WaitForDeploymentAvailableInterval []interface{}
+}
+
+func DeployUnevictablePod(ctx context.Context, input DeployUnevictablePodInput) {
+	workloadClient := input.WorkloadClusterProxy.GetClientSet()
+	numberOfDeployments := 3
+	if input.ControlPlane != nil {
+		untaintMasterNodesAndAddNodeSelectorToPod(ctx, untaintMasterNodesAndAddNodeSelectorToPodInput{
```
Comment (fabriziopandini) [inline test/framework/deployment_helpers.go:298]: Instead of untainting the nodes, what about adding a nodeselector selector targeting nodes with "node-role.kubernetes.io/master" and a toleration on the machine deployment?

the toleration should be similar to the one kubeadm adds to core DNS:

```
      tolerations:
      - key: CriticalAddonsOnly
        operator: Exists
      - key: node-role.kubernetes.io/master
        effect: NoSchedule
```

File: test/e2e/node_drain_timeout.go [hunk 22]
```diff
@@ -0,0 +1,154 @@
...
+				WaitForMachineDeployments: nodeDrainTimeoutMachineDeploymentInterval,
+				Replicas:                  0,
+			})
+		}
+
+		By("Deploy workload on the master node. The workload is actually the pods that we deployed above.")
+		framework.DeployUnevictablePod(ctx, framework.DeployUnevictablePodInput{
+			WorkloadClusterProxy:               workloadClusterProxy,
+			ControlPlane:                       controlplane,
+			IsDeploymentExisted:                true,
```
Comment (fabriziopandini) [inline test/e2e/node_drain_timeout.go:124]: This confuses me (and sorry of I was not clear enough in my comment) 
I think that it will be much simpler to read and maintain if, at this stage we create a new deployment with the required tolerations and node selector (in stead of modifying the existing one)

Comment (namnx228) [reply test/e2e/node_drain_timeout.go]: @fabriziopandini. Yes, i have changed the PR that now we have two separated deployments for testing machine deployment and KCP. Please take a look

File: test/framework/deployment_helpers.go (hunk 21 above)
Comment (namnx228) [reply test/framework/deployment_helpers.go]: Fixed. Thanks

**Comment by fabriziopandini** [comment]
@namnx228 thanks for your patience and huge work in addressing comments!
/lgtm 
/test pull-cluster-api-e2e-full-main

@wfernandes PTAL whenever you have some time

**Comment by wfernandes** [comment]
If its ok, I'll review this tomorrow. Thanks.

File: test/e2e/node_drain_timeout.go [hunk 23]
```diff
@@ -0,0 +1,155 @@
...
+			framework.ScaleAndWaitMachineDeployment(ctx, framework.ScaleAndWaitMachineDeploymentInput{
+				ClusterProxy:              input.BootstrapClusterProxy,
+				Cluster:                   cluster,
+				MachineDeployment:         md,
+				WaitForMachineDeployments: nodeDrainTimeoutMachineDeploymentInterval,
+				Replicas:                  0,
+			})
+		}
+
+		By("Deploy workload on the master node. The workload is actually the pods that we deployed above.")
```
Comment (wfernandes) [inline test/e2e/node_drain_timeout.go:120]: nit: naming - avoid references to "master"
```suggestion
		By("Deploy deployment with unevictable pod on control plane nodes.")
```

File: test/e2e/node_drain_timeout.go [hunk 24]
```diff
@@ -0,0 +1,155 @@
...
+		})
+		cluster = applyClusterTemplateResult.Cluster
+		controlplane = applyClusterTemplateResult.ControlPlane
+		machineDeployments = applyClusterTemplateResult.MachineDeployments
+
+		By("Add a deployment and podDisruptionBudget to the workload cluster. The deployed pods cannot be evicted in the node draining process.")
+		workloadClusterProxy := input.BootstrapClusterProxy.GetWorkloadCluster(context.TODO(), cluster.Namespace, cluster.Name)
+		framework.DeployUnevictablePod(ctx, framework.DeployUnevictablePodInput{
+			WorkloadClusterProxy:               workloadClusterProxy,
+			ControlPlane:                       nil,
```
Comment (wfernandes) [inline test/e2e/node_drain_timeout.go:102]:
```suggestion
```

File: test/e2e/node_drain_timeout.go [hunk 25]
```diff
@@ -0,0 +1,155 @@
...
+			},
+			WaitForClusterIntervals:      input.E2EConfig.GetIntervals(specName, "wait-cluster"),
+			WaitForControlPlaneIntervals: input.E2EConfig.GetIntervals(specName, "wait-control-plane"),
+			WaitForMachineDeployments:    input.E2EConfig.GetIntervals(specName, "wait-worker-nodes"),
+		})
+		cluster = applyClusterTemplateResult.Cluster
+		controlplane = applyClusterTemplateResult.ControlPlane
+		machineDeployments = applyClusterTemplateResult.MachineDeployments
+
+		By("Add a deployment and podDisruptionBudget to the workload cluster. The deployed pods cannot be evicted in the node draining process.")
```
Comment (wfernandes) [inline test/e2e/node_drain_timeout.go:98]:
```suggestion
		By("Add a deployment with unevictable pod and podDisruptionBudget to the workload cluster. The deployed pods cannot be evicted in the node draining process.")
```

File: test/framework/deployment_helpers.go [hunk 26]
```diff
@@ -245,3 +283,124 @@ func WaitForDNSUpgrade(ctx context.Context, input WaitForDNSUpgradeInput, interv
...
+	DeploymentName                     string
+	WaitForDeploymentAvailableInterval []interface{}
+}
+
+func DeployUnevictablePod(ctx context.Context, input DeployUnevictablePodInput) {
+	workloadClient := input.WorkloadClusterProxy.GetClientSet()
+	workloadDeployment := &appsv1.Deployment{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      input.DeploymentName,
+			Namespace: "default",
```
Comment (wfernandes) [inline test/framework/deployment_helpers.go:299]: nit: I wonder if we can use the namespace that was created as part of the spec setup instead of "default" in order to avoid any cross test conflicts later down the road.

**Review by wfernandes — COMMENTED** [review]
@namnx228 Thanks for this PR. This mostly LGTM, just a few nits and a function cleanup.

**Comment by wfernandes** [comment]
@namnx228 Thanks for updating the PR with the fixes! This LGTM. Feel free to squash the commits. 🙂

File: test/framework/deployment_helpers.go [hunk 27]
```diff
@@ -245,3 +250,139 @@ func WaitForDNSUpgrade(ctx context.Context, input WaitForDNSUpgradeInput, interv
...
+
+func DeployUnevictablePod(ctx context.Context, input DeployUnevictablePodInput) {
+	Expect(input.DeploymentName).ToNot(BeNil(), "Need a deployment name in DeployUnevictablePod")
+	Expect(input.Namespace).ToNot(BeNil(), "Need a namespace in DeployUnevictablePod")
+	Expect(input.WorkloadClusterProxy).ToNot(BeNil(), "Need a workloadClusterProxy in DeployUnevictablePod")
+	workloadClient := input.WorkloadClusterProxy.GetClientSet()
+
+	log.Logf("Check if namespace %s exists", input.Namespace)
+	if _, err := workloadClient.CoreV1().Namespaces().Get(ctx, input.Namespace, metav1.GetOptions{}); err != nil {
+		workloadClient.CoreV1().Namespaces().Create(ctx, &corev1.Namespace{
```
Comment (wfernandes) [inline test/framework/deployment_helpers.go:271]: Seems like from the CI failure, we need to check the err here.

Comment (namnx228) [reply test/framework/deployment_helpers.go]: Yes, I have fixed it. Thanks

**Comment by namnx228** [comment]
Hi @wfernandes @fabriziopandini All nits and issue have been solved. Please take a look again. Thanks

**Comment by wfernandes** [comment]
/lgtm
@vincepri for approval

**Review by vincepri — APPROVED** [review]
/approve

**Comment by vincepri** [comment]
Thanks @namnx228 for your great work here!

**Comment by namnx228** [comment]
@fabriziopandini @wfernandes I have rebased my PR. Please give a lgtm again.

**Comment by namnx228** [comment]
Thanks @fabriziopandini @wfernandes @sedefsavas for your reviews. Thanks @vincepri for your approval.
